sale_procurement_group_by_line/tests/test_sale_procurement_group_by_line.py

```diff
@@ -0,0 +1,86 @@
+        self.product_ctg_model = self.env['product.category']
+        self.sale_model = self.env['sale.order']
+        self.order_line_model = self.env['sale.order.line']
+        # Customer
+        self.customer = self.env.ref('base.res_partner_2')
+        # Warehouse
+        self.warehouse_id = self.env.ref('stock.warehouse0')
+        # Create product category
+        self.product_ctg = self._create_product_category()
+        # Create Products
+        self.new_product1 = self._create_product('test_product1')
+        self.new_product2 = self._create_product('test_product2')
+        self.sale = self._create_sale_order()
+
+    def _create_product_category(self):
+        product_ctg = self.product_ctg_model.create({
+            'name': 'test_product_ctg',
+        })
+        return product_ctg
+
+    def _create_product(self, name):
+        product = self.product_model.create({
+            'name': name,
+            'categ_id': self.product_ctg.id,
+            'type': 'product',
+        })
+        return product
+
+    def _create_sale_order(self):
+        """Create a Sale Order."""
+        self.sale = self.sale_model.create({
+            'partner_id': self.customer.id,
+            'warehouse_id': self.warehouse_id.id,
+            'picking_policy': 'direct',
+        })
+        self.line1 = self.order_line_model.create({
+            'order_id': self.sale.id,
+            'product_id': self.new_product1.id,
+            'product_uom_qty': 10.0,
+            'name': 'Sale Order Line Demo1',
+        })
+        self.line2 = self.order_line_model.create({
+            'order_id': self.sale.id,
+            'product_id': self.new_product2.id,
+            'product_uom_qty': 5.0,
+            'name': 'Sale Order Line Demo2',
+        })
+        return self.sale
+
+    def test_procurement_group_by_line(self):
+        self.sale.action_confirm()
+        self.assertEqual(self.line2.procurement_group_id,
+                         self.line1.procurement_group_id,
+                         """Both Sale Order line should belong
+                         to Procurement Group""")
+        self.picking_ids = self.env['stock.picking'].\
+            search([('group_id', 'in', self.line2.procurement_group_id.ids)])
+        self.picking_ids.move_lines.write({'quantity_done': 5})
+        wiz_act = self.picking_ids.button_validate()
+        wiz = self.env[wiz_act['res_model']].browse(wiz_act['res_id'])
+        wiz.process()
+        self.assertTrue(self.picking_ids,
+                        'Procurement Group should have picking')
+
+    def test_action_launch_procurement_rule(self):
+        group_id = self.env['procurement.group'].create(
+            {'move_type': 'one',
+             'sale_id': self.sale.id,
+             'name': self.sale.name})
+        self.line1.procurement_group_id = group_id
+        self.line2.procurement_group_id = group_id
+        self.sale.action_confirm()
```

Comment:
no assert in this test?

Reply:
and what we should assert?

Reply:
the expected behavior? :laughing:  If you understand the module you should be able to know what to assert.

Reply:
but that's not constructive 😒

Reply:
¬¬ IMHO just checking two procurement groups are created is ok. Perhaps also the names for the procurment groups...

Reply:
a lot of asserts implemented :)